gotwebctl/gotwebctl.8
.\"
.\" Copyright (c) 2026 Stefan Sperling
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt GOTWEBCTL 8
.Os
.Sh NAME
.Nm gotwebctl
.Nd control the Game of Trees Web Daemon
.Sh SYNOPSIS
.Nm
.Op Fl hV
.Op Fl f Ar path
.Ar command
.Op Ar arg ...
.Sh DESCRIPTION
.Nm
controls the
.Xr gotwebd 8
daemon.
.Pp
Access to the
.Xr gotwebd 8
control socket requires root privileges.
.Pp
The options for
.Nm
are as follows:
.Bl -tag -width Ds
.It Fl h
Display usage information and exit immediately.
.It Fl f Ar path
Set the
.Ar path
to the control socket which
.Xr gotwebd 8
is listening on.
If not specified, the default path
.Pa /var/run/gotwebd.sock
will be used.
.It Fl V , -version
Display program version and exit immediately.
.El
.Pp
The commands for
.Nm
are as follows:
.Bl -tag -width Ds
.It Cm info
Display information about a running
.Xr gotwebd 8
instance.
.It Cm stop
Ask a running
.Xr gotwebd 8
instance to gracefully stop itself.
This is equivalent to sending a TERM signal to
.Xr gotwebd 8
without requiring knowledge of the process ID to send the signal to.
.Pp
The control socket will be closed by
.Xr gotwebd 8 ,
allowing a new instance of
.Xr gotwebd 8
to take over.
New FastCGI socket connections will no longer be accepted, but FastCGI
requests currently in progress are still processed to completion.
.Sh SEE ALSO
.Xr got 1 ,
.Xr gotwebd.conf 5 ,
.Xr gotwebd 8
.Sh AUTHORS
.An Stefan Sperling Aq Mt stsp@openbsd.org


CHANGES
* got 0.120; 2025-10-03
- disable gotwebd authentication if it is not enabled in /etc/gotwebd.conf
- ensure that GOTWEBD_LOGIN_TIMEOUT is used consistently at build time
- prevent date-specific gotsysd regress failures due to asctime_r whitespace
- make gotwebd refuse to start up if the _gotwebd user is root
- make gotwebd warn if the webserver's user is set to root in /etc/gotwebd.conf
- add /etc/gotwebd.conf parameters for hiding repositories
- reject bad hostnames provided to the gotsh weblogin command
- allow gotwebd to optionally display a login hint when authentication fails

* got 0.119; 2025-09-29
- clean up gotwebd requests before freeing sockets; avoids deref of NULL sock
- fix wrong process names in gotwebd error messages
- use an absolute URL in redirects generated by gotwebd
- fix an issue where histedit would not merge added files correctly
- tweak gotwebd deferred accept handling to work more like httpd's
- add a 'weblogin' command to gotsh for use with gotwebd user authentication
- add user authentication support to gotwebd
- make gotwebd IPC sockets use non-blocking I/O as intended to fix hang
- prevent gotd looping on unhandled errors raised by accept()
- fix gotsysd.conf "listen on" statement failing on non-existent sockets
- pass gotwebd.conf's www user setting to every gotwebd child process

* got 0.118; 2025-09-09
- security fix for -portable: gotwebd can be tricked into reading repositories
  outside its repos_path; bug introduced in got-0.111; OpenBSD is not affected
- make 'tog diff' show the repository name in names of patches written to /tmp
- plug memory leaks which were making gotwebd regress tests fail
- fix parallel processing of requests in gotwebd, improving responsiveness
- set gotwebd pledges according to address families of listening sockets
- run gotwebd fcgi parameter parsing in a dedicated process under pledge "stdio"
- make gotd commit notifications only show history which is unique to the branch
- enable sftp/scp support in the sshd_config file generated by gotsysd
- make gotsysd-managed repositories readable for the _gotd group

* got 0.117; 2025-08-17
- regress: replace "sed -i" with ed(1) for portable in-place editing
- ensure that error messages from gotsysd libexec helpers get logged
- fix gotsysd using wrong auth and hmac labels in the generated gotd.conf
- preserve bad symlinks across merges during rebase and histedit
- improve binary files detection: detect any control characters, not just NUL
- gotwebd: fix race condition resulting in trucated html with trailing garbage
- make commit coloring faster and more accurate, producing smaller pack files
- improve selection of pack files for pinning in the open pack file cache
- regress: don't load global/home git configuration files while running tests
- make 'got clone' set a got.conf default branch for fetching only, not sending

* got 0.116; 2025-07-25
- make our pack-refs header format align with the expectations of git 2.50.0
- fix bogus "bad offset in pack file" errors wrongly raised by gotd
- fix gotd branch protection rejecting commits that already exist on server
- pick a default branch to clone when the server does not advertise HEAD symref
- do not clobber changes staged via stage -p during "got revert"
- enforce additional restrictions on reference names specified in gotsys.conf
- change gotwebd favicons to show the smiley fish only
- fix gotd reload when /etc/gotd-secrets.conf is used
- fix bogus "raw object has unexpected size" errors during deltification
- fix bug in delta block stretch size calculation resulting in invalid deltas
- fix gotsysd behaviour when the anonymous user is removed from gotsys.conf
- add support for email and http/json notifications to gotsysd and gotsys.conf

* got 0.115; 2025-06-27
- make errors reported by gotsys-apply-conf actually visible
- stop trying to start gotd from gotsys-apply-conf if gotd is not running
- fix infinite loop in got_pack_repaint_parent_commits() and got-read-pack
- fix creation of gotd.conf deny rules in gotsys-write-conf
- add support for global repository access rules to gotsysd.conf
- fix segfault due to double-free in got-read-gotconfig

* got 0.114; 2025-06-23
- preserve author timestamps when rebasing commits
- stop running ssh with -q by default; -q hides host key fingerprint errors
- fix gotsys-read-conf crash when ssh key comments are missing in gotsys.conf
- relax repository path permission checks in gotsys-repo-create
- add gotsys apply -w option which waits until sysconf has been run
- fix gotsysd getting stuck due to missing final messages from libexec helpers
- plug a file descriptor leak in the gotsysd libexec process

* got 0.113; 2025-06-05
- tweak 'got status' and 'got add' ignores handling for better git compatibility
- improve redundant pack detection during 'gotadmin cleanup'
- gotwebd: do not forget to initialize *ngroups argument for getgrouplist()
- fix default access for root and _gotd when gotsysd runs without config file
- fix bogus "unexpected privsep message" from gotsh during 'got send'
- fix a race in gotd notification processing causing notify process to exit

* got 0.112; 2025-05-14
- remove /tmp/got-importmsg temp files when import commit message is left empty
- rely on secondary _gotwebd groups if repos_path is not owned by _gotwebd group
- fix unrelated errors being reported if a histedit operation is aborted
- implement support for protected references in gotsys.conf and gotsysd
- plug memory leaks in some libexec helpers and in the gitconfig parser
- stop needlessly opening the repository whenever a work tree is opened

* got 0.111; 2025-04-22
- introduce gotsysd: configure gotd servers by committing to gotsys.git repo
- make gotd run 'gotsys check' on gotsys.conf commits before accepting them
- make gotd run 'gotsys apply' when the gotsys.git repo receives changes
- add a missing malloc failure check to gotd's repo_write process

gotwebctl/gotwebctl.c
/*
 * Copyright (c) 2026 Stefan Sperling <stsp@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/queue.h>
#include <sys/tree.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/stat.h>

#include <err.h>
#include <errno.h>
#include <event.h>
#include <fcntl.h>
#include <imsg.h>
#include <limits.h>
#include <locale.h>
#include <sha1.h>
#include <sha2.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <getopt.h>
#include <unistd.h>

#include "got_error.h"
#include "got_object.h"
#include "got_version.h"
#include "got_path.h"
#include "got_reference.h"

#include "media.h"
#include "gotwebd.h"

#ifndef nitems
#define nitems(_a)	(sizeof((_a)) / sizeof((_a)[0]))
#endif

#define GOTCTL_CMD_INFO "info"
#define GOTCTL_CMD_STOP "stop"

struct gotwebctl_cmd {
	const char	*cmd_name;
	const struct got_error *(*cmd_main)(int, char *[], int);
	void		(*cmd_usage)(int);
};

__dead static void	usage(int, int);

__dead static void	usage_info(int);
__dead static void	usage_stop(int);

static const struct got_error*		cmd_info(int, char *[], int);
static const struct got_error*		cmd_stop(int, char *[], int);

static const struct gotwebctl_cmd gotwebctl_commands[] = {
	{ "info",	cmd_info,	usage_info },
	{ "stop",	cmd_stop,	usage_stop },
};

__dead static void
usage_info(int status)
{
	FILE *fp = (status == 0) ? stdout : stderr;
	fprintf(fp, "usage: %s info\n", getprogname());
	exit(status);
}

static const struct got_error *
show_info(struct imsg *imsg)
{
	struct gotwebd_imsg_info info;
	size_t datalen;

	datalen = imsg->hdr.len - IMSG_HEADER_SIZE;
	if (datalen != sizeof(info))
		return got_error(GOT_ERR_PRIVSEP_LEN);
	memcpy(&info, imsg->data, sizeof(info));

	printf("gotwebd PID: %d\n", info.pid);
	printf("verbosity: %d\n", info.verbosity);
	return NULL;
}

static const struct got_error *
cmd_info(int argc, char *argv[], int gotwebd_sock)
{
	const struct got_error *err = NULL;
	struct imsgbuf ibuf;
	struct imsg imsg;
	ssize_t n;
	int done = 0;

	if (unveil(NULL, NULL) != 0)
		return got_error_from_errno("unveil");
#ifndef PROFILE
	if (pledge("stdio", NULL) == -1)
		return got_error_from_errno("pledge");
#endif
	if (imsgbuf_init(&ibuf, gotwebd_sock) == -1)
		return got_error_from_errno("imsgbuf_init");

	if (imsg_compose(&ibuf, GOTWEBD_IMSG_CTL_INFO, 0, 0, -1,
	    NULL, 0) == -1) {
		imsgbuf_clear(&ibuf);
		return got_error_from_errno("imsg_compose INFO");
	}

	if (imsgbuf_flush(&ibuf) == -1) {
		imsgbuf_clear(&ibuf);
		return got_error_from_errno("imsgbuf_flush");
	}

	while (!done && err == NULL) {
		n = imsgbuf_read(&ibuf);
		if (n == -1) {
			if  (errno != EAGAIN) {
				err = got_error_from_errno("imsgbuf_read");
				break;
			}
				
			sleep(1);
			continue;
		}
		if (n == 0)
			break;

		n = imsg_get(&ibuf, &imsg);
		if (n == -1) {
			err = got_error_from_errno("imsg_get");
			break;
		}

		if (n == 0)
			break;

		switch (imsg.hdr.type) {
		case GOTWEBD_IMSG_CTL_INFO:
			err = show_info(&imsg);
			done = 1;
			break;
		default:
			err = got_error(GOT_ERR_PRIVSEP_MSG);
			break;
		}

		imsg_free(&imsg);
	}

	imsgbuf_clear(&ibuf);
	return err;
}

__dead static void
usage_stop(int status)
{
	FILE *fp = (status == 0) ? stdout : stderr;
	fprintf(fp, "usage: %s stop\n", getprogname());
	exit(status);
}

static const struct got_error *
cmd_stop(int argc, char *argv[], int gotwebd_sock)
{
	const struct got_error *err;
	struct imsgbuf ibuf;
	struct imsg imsg;
	ssize_t n;

	if (unveil(NULL, NULL) != 0)
		return got_error_from_errno("unveil");
#ifndef PROFILE
	if (pledge("stdio", NULL) == -1)
		return got_error_from_errno("pledge");
#endif
	if (imsgbuf_init(&ibuf, gotwebd_sock) == -1)
		return got_error_from_errno("imsgbuf_init");

	if (imsg_compose(&ibuf, GOTWEBD_IMSG_CTL_STOP, 0, 0, -1,
	    NULL, 0) == -1) {
		imsgbuf_clear(&ibuf);
		return got_error_from_errno("imsg_compose STOP");
	}

	if (imsgbuf_flush(&ibuf) == -1) {
		imsgbuf_clear(&ibuf);
		return got_error_from_errno("imsgbuf_flush");
	}

	for (;;) {
		n = imsg_get(&ibuf, &imsg);
		if (n == -1) {
			err = got_error_from_errno("imsg_get");
			break;
		}

		if (n == 0)
			break;

		switch (imsg.hdr.type) {
		default:
			err = got_error(GOT_ERR_PRIVSEP_MSG);
			break;
		}

		imsg_free(&imsg);
	}

	imsgbuf_clear(&ibuf);
	return err;
}

static void
list_commands(FILE *fp)
{
	size_t i;

	fprintf(fp, "commands:");
	for (i = 0; i < nitems(gotwebctl_commands); i++) {
		const struct gotwebctl_cmd *cmd = &gotwebctl_commands[i];
		fprintf(fp, " %s", cmd->cmd_name);
	}
	fputc('\n', fp);
}

__dead static void
usage(int hflag, int status)
{
	FILE *fp = (status == 0) ? stdout : stderr;

	fprintf(fp, "usage: %s [-hV] [-f path] command [arg ...]\n",
	    getprogname());
	if (hflag)
		list_commands(fp);
	exit(status);
}

static int
connect_gotwebd(const char *socket_path)
{
	int gotwebd_sock = -1;
	struct sockaddr_un sun;

	if (unveil(socket_path, "w") != 0)
		err(1, "unveil %s", socket_path);

	if ((gotwebd_sock = socket(AF_UNIX, SOCK_STREAM, 0)) == -1)
		err(1, "socket");

	memset(&sun, 0, sizeof(sun));
	sun.sun_family = AF_UNIX;
	if (strlcpy(sun.sun_path, socket_path, sizeof(sun.sun_path)) >=
	    sizeof(sun.sun_path))
		errx(1, "gotd socket path too long");
	if (connect(gotwebd_sock, (struct sockaddr *)&sun, sizeof(sun)) == -1)
		err(1, "connect: %s", socket_path);

	return gotwebd_sock;
}

int
main(int argc, char *argv[])
{
	const struct gotwebctl_cmd *cmd;
	int gotwebd_sock = -1, i;
	int ch;
	int hflag = 0, Vflag = 0;
	static const struct option longopts[] = {
	    { "version", no_argument, NULL, 'V' },
	    { NULL, 0, NULL, 0 }
	};
	const char *socket_path = GOTWEBD_CONTROL_SOCKET;

	setlocale(LC_CTYPE, "");

#ifndef PROFILE
	if (pledge("stdio rpath unix unveil", NULL) == -1)
		err(1, "pledge");
#endif
	while ((ch = getopt_long(argc, argv, "+hf:V", longopts, NULL)) != -1) {
		switch (ch) {
		case 'h':
			hflag = 1;
			break;
		case 'f':
			socket_path = optarg;
			break;
		case 'V':
			Vflag = 1;
			break;
		default:
			usage(hflag, 1);
			/* NOTREACHED */
		}
	}

	argc -= optind;
	argv += optind;
	optind = 1;
	optreset = 1;

	if (Vflag) {
		got_version_print_str();
		return 0;
	}

	if (argc <= 0)
		usage(hflag, hflag ? 0 : 1);

	for (i = 0; i < nitems(gotwebctl_commands); i++) {
		const struct got_error *error;

		cmd = &gotwebctl_commands[i];

		if (strncmp(cmd->cmd_name, argv[0], strlen(argv[0])) != 0)
			continue;

		if (hflag)
			cmd->cmd_usage(0);
#ifdef PROFILE
		if (unveil("gmon.out", "rwc") != 0)
			err(1, "unveil", "gmon.out");
#endif
		gotwebd_sock = connect_gotwebd(socket_path);
		if (gotwebd_sock == -1)
			return 1;
		error = cmd->cmd_main(argc, argv, gotwebd_sock);
		close(gotwebd_sock);
		if (error && error->msg[0] != '\0') {
			fprintf(stderr, "%s: %s\n", getprogname(), error->msg);
			return 1;
		}

		return 0;
	}

	fprintf(stderr, "%s: unknown command '%s'\n", getprogname(), argv[0]);
	list_commands(stderr);
	return 1;
}
